src/main/java/io/vertx/core/net/impl/pool/CombinerExecutor.java

```diff
@@ -12,61 +12,149 @@
 
 import io.netty.util.internal.PlatformDependent;
 
+import java.time.Duration;
+import java.util.Objects;
 import java.util.Queue;
+import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.LongSupplier;
 
 /**
  * Lock free executor.
- *
+ * <p>
  * When a thread submits an action, it will enqueue the action to execute and then try to acquire
  * a lock. When the lock is acquired it will execute all the tasks in the queue until empty and then
  * release the lock.
  */
 public class CombinerExecutor<S> implements Executor<S> {
+  @FunctionalInterface
+  public interface YieldCondition {
+    boolean yield(long actions);
+
+    static YieldCondition withMaxDuration(final Duration duration) {
+      return withMaxDuration(duration, System::nanoTime);
+    }
+
+    static YieldCondition withMaxDuration(final Duration duration, final LongSupplier nanoTimeSupplier) {
+      return new YieldCondition() {
+        private final long durationInNs = Objects.requireNonNull(duration).toNanos();
+        private long start;
+
+        @Override
+        public boolean yield(long actions) {
+          assert actions >= 0;
+          if (actions == 0) {
+            start = nanoTimeSupplier.getAsLong();
+            return false;
+          }
+          final long elapsed = nanoTimeSupplier.getAsLong() - start;
+          assert elapsed >= 0;
+          return elapsed >= durationInNs;
+        }
+      };
+    }
+
+    static YieldCondition withMaxCount(final int maxActions) {
+      if (maxActions < 0) {
+        throw new IllegalArgumentException("maxActions must be >= 1");
+      }
+      return count -> count >= maxActions;
+    }
+  }
 
   private final Queue<Action<S>> q = PlatformDependent.newMpscQueue();
   private final AtomicInteger s = new AtomicInteger();
   private final S state;
 
-  public CombinerExecutor(S state) {
+  private final YieldCondition yieldCondition;
```

Comment:
I think the yieldCondition should be an argument of the submitAndContinue instead of being a field.

Reply:
Good idea, less states inside the executor

Reply:
I think it's fine keeping it like this, so we don't need to deal with potentially different conditions that would be randomly used